lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -336,17 +337,32 @@ def set_aspect(self, aspect, adjustable=None, anchor=None, share=False):
             view_intervals = np.array([self.xaxis.get_view_interval(),
                                        self.yaxis.get_view_interval(),
                                        self.zaxis.get_view_interval()])
-            mean = np.mean(view_intervals, axis=1)
-            ptp = np.ptp(view_intervals, axis=1)
-            delta = max(ptp[ax_indices])
-            scale = self._box_aspect[ptp == delta][0]
-            deltas = delta * self._box_aspect / scale
-
-            for i, set_lim in enumerate((self.set_xlim3d,
-                                         self.set_ylim3d,
-                                         self.set_zlim3d)):
-                if i in ax_indices:
-                    set_lim(mean[i] - deltas[i]/2., mean[i] + deltas[i]/2.)
+            if adjustable == 'datalim':
+                mean = np.mean(view_intervals, axis=1)
+                ptp = np.ptp(view_intervals, axis=1)
+                delta = max(ptp[ax_indices])
+                scale = self._box_aspect[ptp == delta][0]
+                deltas = delta * self._box_aspect / scale
+
+                for i, set_lim in enumerate((self.set_xlim3d,
+                                             self.set_ylim3d,
+                                             self.set_zlim3d)):
+                    if i in ax_indices:
+                        set_lim(mean[i] - deltas[i]/2., mean[i] + deltas[i]/2.)
+            elif adjustable == 'box':
+                # Change the box aspect such that the ratio of the length of
+                # the unmodified axis to the length of the diagonal
+                # perpendicular to it remains unchanged.
+                deltas = np.ptp(view_intervals, axis=1)
+                box_aspect = np.array(self.get_box_aspect())
+                box_aspect[ax_indices] = deltas[ax_indices]
+                remaining_ax_indices = {0, 1, 2}.difference(ax_indices)
```

Comment:
This can probably go first, and the top three lines inside the if-clause (your approach is still much better than my original...).

Reply:
A little confused here.:sweat_smile: Could you explain a bit more where it can go? (I thought it's better to avoid calculating the norm if it's not required?)

Reply:
Ahh, I remembered incorrectly I thing. I had the idea that we do not need to update `box_aspect` unless remaining_ax_indices was non-empty. But I realize I was wrong.